### ISSUE TITLE
"Docker version 17.12.0 or later required." with Docker Engine 24.0.7

### ISSUE THREAD

**Issue by schauerte** [microsoft/vscode-remote-release#10027]
### Problem

While trying to use a locally installed docker engine to run a DevContainer, I'm falsely requested to install an older version to start working.

### Environment
- VSCode Version: 1.90.2
- Local OS Version: Ubuntu 22.04 (TuxedoOS)
- Remote OS Version: - 
- Remote Extension/Connection Type: DevContainers

### Logs
"Dev Containers Developer: Show All Logs" latest (from 4th of July):
  > [2024-07-04T11:24:14.930Z] Dev Containers 0.369.0 in VS Code 1.90.2 (5437499feb04f7a586f677b155b039bc2b3669eb).
[2024-07-04T11:24:14.930Z] Initializing configuration support...
[2024-07-04T11:24:14.930Z] Internal initialization of dev container support package...
[2024-07-04T11:24:16.166Z] Start: Run: docker version --format {{json .}}
[2024-07-04T11:24:16.169Z] Stop (3 ms): Run: docker version --format {{json .}}
[2024-07-04T11:24:16.170Z] Stop (4 ms): Run: docker version --format {{json .}}
[2024-07-04T11:24:16.170Z] CLI host's PATH: /app/bin:/app/bin:/app/bin:/usr/bin:/home/localuser/.var/app/com.visualstudio.code/data/node_modules/bin

The same with Pre-Release (also 4th of July): 
> Dev Containers 0.376.0 in VS Code 1.90.2 (5437499feb04f7a586f677b155b039bc2b3669eb).

"Dev Containers: Show container Log" (from 7th of July - I don't know why the version differs):
> [93 ms] Dev Containers 0.375.0 in VS Code 1.90.2 (5437499feb04f7a586f677b155b039bc2b3669eb).
[92 ms] Initializing configuration support...
[93 ms] Internal initialization of dev container support package...
[1432 ms] Start: Run: docker version --format {{json .}}
[1438 ms] CLI host's PATH: /app/bin:/app/bin:/app/bin:/usr/bin:/home/andreas/.var/app/com.visualstudio.code/data/node_modules/bin

### Steps to Reproduce

1. `code .` in a local folder with `.devcontainer/devcontainer.json`
2. *Reopen in container*
3. MessageBox "Docker version 17.12.0 or later required." is displayed (with options: License, Cancel and Install)

### Additional Information

Does this issue occur when you try this locally?: Yes
Does this issue occur when you try this locally and all extensions are disabled?: yes (with everything disabled but *Dev Containers*)

`docker -v`:
> Docker version 24.0.7, build 24.0.7-0ubuntu2~22.04.1

`docker version --format '{{json .}}'`:
> `{"Client":{"Version":"24.0.7","ApiVersion":"1.43","DefaultAPIVersion":"1.43","GitCommit":"24.0.7-0ubuntu2~22.04.1","GoVersion":"go1.21.1","Os":"linux","Arch":"amd64","BuildTime":"Wed Mar 13 20:23:54 2024","Context":"default"},"Server":{"Platform":{"Name":""},"Components":[{"Name":"Engine","Version":"24.0.7","Details":{"ApiVersion":"1.43","Arch":"amd64","BuildTime":"Wed Mar 13 20:23:54 2024","Experimental":"false","GitCommit":"24.0.7-0ubuntu2~22.04.1","GoVersion":"go1.21.1","KernelVersion":"6.5.0-10043-tuxedo","MinAPIVersion":"1.12","Os":"linux"}},{"Name":"containerd","Version":"1.7.12","Details":{"GitCommit":""}},{"Name":"runc","Version":"1.1.12-0ubuntu2~22.04.1","Details":{"GitCommit":""}},{"Name":"docker-init","Version":"0.19.0","Details":{"GitCommit":""}}],"Version":"24.0.7","ApiVersion":"1.43","MinAPIVersion":"1.12","GitCommit":"24.0.7-0ubuntu2~22.04.1","GoVersion":"go1.21.1","Os":"linux","Arch":"amd64","KernelVersion":"6.5.0-10043-tuxedo","BuildTime":"2024-03-13T20:23:54.000000000+00:00"}}`

`docker version --format {{json .}}`
In zsh: 
> zsh: parse error near `}'

In bash:
> "docker version" accepts no arguments.

**Comment by schauerte** [comment]
BTW: I can't find the extensions source code. Isn't it open sourced?

**Comment by chrmarti** [comment]
Please append the Dev Containers log from when this happens. (`F1` > `Dev Containers: Show Container Log`)

**Comment by schauerte** [comment]
> Please append the Dev Containers log from when this happens. (`F1` > `Dev Containers: Show Container Log`)

Added to the description above.

**Comment by pomeh** [comment]
Hello,

I encounter the same error "Docker version 17.12.0 or later required." since today, was ok until now.

OS: MacOS Version 13.6.7

```
$ code --version
1.91.0
ea1445cc7016315d0f5728f8e8b12a45dc0a7286
x64

$ docker -v
Docker version 26.1.4, build 5650f9b

$ docker version --format '{{json .}}'
{"Client":{"Version":"26.1.4","ApiVersion":"1.45","DefaultAPIVersion":"1.45","GitCommit":"5650f9b","GoVersion":"go1.21.11","Os":"darwin","Arch":"amd64","BuildTime":"Wed Jun  5 11:26:02 2024","Context":"desktop-linux"},"Server":{"Platform":{"Name":"Docker Desktop 4.31.0 (153195)"},"Components":[{"Name":"Engine","Version":"26.1.4","Details":{"ApiVersion":"1.45","Arch":"amd64","BuildTime":"Wed Jun  5 11:29:22 2024","Experimental":"false","GitCommit":"de5c9cf","GoVersion":"go1.21.11","KernelVersion":"6.6.31-linuxkit","MinAPIVersion":"1.24","Os":"linux"}},{"Name":"containerd","Version":"1.6.33","Details":{"GitCommit":"d2d58213f83a351ca8f528a95fbd145f5654e957"}},{"Name":"runc","Version":"1.1.12","Details":{"GitCommit":"v1.1.12-0-g51d5e94"}},{"Name":"docker-init","Version":"0.19.0","Details":{"GitCommit":"de40ad0"}}],"Version":"26.1.4","ApiVersion":"1.45","MinAPIVersion":"1.24","GitCommit":"de5c9cf","GoVersion":"go1.21.11","Os":"linux","Arch":"amd64","KernelVersion":"6.6.31-linuxkit","BuildTime":"2024-06-05T11:29:22.000000000+00:00"}}
```

`Dev Containers: Show Container Log`:

```
[725 ms] Dev Containers 0.375.0 in VS Code 1.91.0 (ea1445cc7016315d0f5728f8e8b12a45dc0a7286).
[724 ms] Initializing configuration support...
[726 ms] Internal initialization of dev container support package...
[80305 ms] Start: Run: docker version --format {{json .}}
[80321 ms] CLI host's PATH: /usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin
[976528 ms] Start: Run: docker version --format {{json .}}
[976538 ms] CLI host's PATH: /usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin
```

-> both in VS Code terminal and in an external iTerm2 terminal, the PATH is slightly different:

```
$ echo $PATH
[...]:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Applications/Docker.app/Contents/Resources/bin/:[...]
```

This `$PATH` includes the docker binary path (`which docker -> /Applications/Docker.app/Contents/Resources/bin//docker`), but the `$PATH` used by Dev Containers does not contains this path.

I've tried [the workaround](https://github.com/microsoft/vscode-remote-release/issues/9105) from [Natielle](https://github.com/Natielle), which explicitely set the Docker path in dev containers settings, but I get another error:

```
ERROR: failed to run Build function: private-registry.example.com/img:v0.0.1: failed to resolve source metadata for private-registry.example.com/img:v0.0.1: error getting credentials - err: exec: "docker-credential-desktop": executable file not found in $PATH, out: ``
```

I'm using a private registry with authentication, and dev container can not authenticate: `docker-credential-desktop executable file not found`, which somewhat confirms me the `$PATH` issue.

Also, since this morning I have another problem with VS Code, when I start a terminal, I get this error:

```
The default interactive shell is now zsh.
To update your account to use zsh, please run `chsh -s /bin/zsh`.
For more details, please visit https://support.apple.com/kb/HT208050.
```

I've never installed nor used zsh on my MacOS, only bash (`GNU bash, version 5.2.26(1)-release (x86_64-apple-darwin22.6.0)`).

Lastly, I got [June 2024 (version 1.91) VS Code update](https://code.visualstudio.com/updates/v1_91) yesterday, which may be the root cause of all this new problems I get, but I can't find anything related to shell/zsh/docker/path/dev container in the [VS Code release notes](https://code.visualstudio.com/updates/v1_91). 

```
Code > "About Visual Studio Code"
Version: 1.91.0
Commit: ea1445cc7016315d0f5728f8e8b12a45dc0a7286
Date: 2024-07-01T18:54:18.541Z (6 days ago)
Electron: 29.4.0
ElectronBuildId: 9728852
Chromium: 122.0.6261.156
Node.js: 20.9.0
V8: 12.2.281.27-electron.0
OS: Darwin x64 22.6.0
```

**Comment by schauerte** [comment]
Thanks @pomeh. I've also tried to specify the path to docker in the settings, but this did not lead to any improvement (it was already in the path listed above).

**Comment by schauerte** [comment]
> I don't like it but... after rebooting the computer, my problem has vanished 😱
> 
> I don't think I've changed anything that could explain both the error I got nor the non-error.

"luckily" rebooting does not change anything for me. @chrmarti cam I provide anything else?

**Comment by chrmarti** [comment]
@schauerte Could you append the full Dev Containers log? I only see snippets in the description.

The output looks good otherwise. Maybe `docker version --format '{{json .}}'` exits with a non-zero exit code?

**Comment by schauerte** [comment]
This is no snippet. It's the full log!

**Comment by schauerte** [comment]
I gave it another thought today and just tried to run the `docker version` command in the integrated terminal. Then I realized that I can't access my `/usr/bin` folder because I have installed VS Code using [flatpak](https://flathub.org/apps/com.visualstudio.code). Fixing the path just leads to a "can't communicate with docker daemon" error. This might be fixed by proper flatpak configuration, but I just switched to the "official" .deb package and everything **works as expected.** 

Sorry for bothering you!

You might want to reopen this issue to improve the logging of problems running the executable.
